java/gateway-core/src/main/java/com/avaulta/gateway/rules/SchemaRuleUtils.java

```diff
@@ -0,0 +1,268 @@
+        return objectMapper.writeValueAsString(filterBySchema(provisionalOutput, schema, schema));
+    }
+
+
+    Object filterBySchema(JsonNode provisionalOutput, JsonSchema schema, JsonSchema root) {
+        if (schema.isRef()) {
+            if (schema.getRef().equals("#")) {
+                // recursive self-reference; see
+                return filterBySchema(provisionalOutput, root, root);
+            } else if (schema.getRef().startsWith("#/definitions/")) {
+                String definitionName = schema.getRef().substring("#/definitions/".length());
+                JsonSchema definition = root.getDefinitions().get(definitionName);
+                return filterBySchema(provisionalOutput, definition, root);
+            } else {
+                //cases like URLs relative to schema URI are not supported
+                throw new RuntimeException("unsupported ref: " + schema.getRef());
+            }
+        } else if (schema.hasType()) {
+            //must have explicit type
+
+            // https://json-schema.org/understanding-json-schema/reference/type.html
+            if (schema.isString()) {
+                if (provisionalOutput.isTextual()) {
+                    //TODO: validate 'format'??
+                    return provisionalOutput.asText();
+                } else {
+                    return null;
+                }
+            } else if (schema.isNumber()) {
+                if (provisionalOutput.isNumber() || provisionalOutput.isNull()) {
+                    return provisionalOutput.numberValue();
+                } else {
+                    return null;
+                }
+            } else if (schema.isInteger()) {
+                if (provisionalOutput.canConvertToInt() || provisionalOutput.isNull()) {
+                    return provisionalOutput.intValue();
+                } else if (provisionalOutput.canConvertToLong()) {
+                    return provisionalOutput.longValue();
+                } else {
+                    return null;
+                }
+            } else if (schema.isBoolean()) {
+                if (provisionalOutput.isBoolean() || provisionalOutput.isNull()) {
+                    return provisionalOutput.booleanValue();
+                } else {
+                    return null;
+                }
+            } else if (schema.isObject()) {
+                if (provisionalOutput.isObject()) {
+                    Map<String, Object> filtered = new HashMap<>();
+                    provisionalOutput.fields().forEachRemaining(entry -> {
+                        String key = entry.getKey();
+                        JsonNode value = entry.getValue();
+                        JsonSchema propertySchema = schema.getProperties().get(key);
+                        if (propertySchema != null) {
+                            Object filteredValue = filterBySchema(value, propertySchema, root);
+                            filtered.put(key, filteredValue);
```

Comment:
we could omit the key if the filtered value is null
```suggestion
                            Optional.ofNullable(filterBySchema(value, propertySchema, root)).ifPresent( v -> filtered.put(key,v));
```